src/main/java/cubicchunks/worldgen/WorldGenerator.java

```diff
@@ -21,260 +21,302 @@
-		CubeCoords next = next();
-		Cube cube = this.cubeProvider.getCube(next);
-		if (cube == null) {
-			return null;
+		// Set the target stage.
+		if (cube.getTargetStage() == null || cube.getTargetStage().precedes(targetStage)) {
+			cube.setTargetStage(targetStage);
 		}
-		GeneratorStage previousStage = cube.getCurrentStage();
-		previousStage.getProcessor().calculate(cube);
 
-		//advance to next stage
-		boolean isLastStage = previousStage.getOrdinal() + 1 >= this.stages.size();
-		GeneratorStage nextStage = isLastStage ?
-				GeneratorStage.LIVE :
-				this.stages.get(previousStage.getOrdinal() + 1);
+		// Generate the cube if necessary (check performed in generateCube)
+		this.generateCube(cube);
 
-		//will be re-registered with new stage in generate()
-		this.dependentCubeManager.unregister(cube);
-		cube.setCurrentStage(nextStage);
-		this.dependentCubeManager.updateDependents(cube);
-		if (!isLastStage && nextStage.precedes(cube.getTargetStage())) {
-			generate(cube);
-		}
-		return previousStage;
+		return cube;
 	}
 
-	private CubeCoords next() {
-		//throws exception when there is no next cube
-		while(true) {
-			CubeCoords coords = this.toUpdate.remove(toUpdate.size() - 1);
-			if(this.toUpdateSet.remove(coords)) {
-				return coords;
-			}
-		}
+	@Override @Nullable
+	public Cube generateCube(@Nonnull CubeCoords coords) {
+		return this.generateCube(coords, GeneratorStage.LIVE);
 	}
 
-	private void sort() {
-		if(this.toUpdate.size() != this.toUpdateSet.size()) {
-			toUpdate.clear();
-			toUpdate.addAll(toUpdateSet);
-		}
-		//TODO: make GeneratorPipeline sort() faster
-		Collection<EntityPlayer> players = world.getPlayerEntities();
-
-		//TODO: is it correct order?
-		Collections.sort(this.toUpdate, (coords1, coords2) ->
-				ComparisonChain.start().compare(
-						getClosestPlayerDistance(coords2, players),
-						getClosestPlayerDistance(coords1, players)
-				).result());
+	@Override @Nullable
+	public Column generateColumn(int cubeX, int cubeZ) {
+		return this.columnGenerator.generateColumn(cubeX, cubeZ);
 	}
 
-	private int getClosestPlayerDistance(CubeCoords coords, Collection<EntityPlayer> players) {
-		int min = Integer.MAX_VALUE;
-		for (EntityPlayer player : players) {
-			int dist = CubeCoords.fromEntity(player).distSquared(coords);
-			if (dist < min) {
-				min = dist;
-			}
-		}
-		return min;
-	}
+	@Override
+	public void calculateAll() {
+
+		long timeStart = System.currentTimeMillis();
 
-	public void generateAll() {
+		int processed = 0;
 		int skipped = 0;
 		int finished = 0;
-		//number of cubes that have reached n-th stage
-		int[] cubesInStage = new int[this.stages.size()];
+		int[] cubesInStage = new int[this.generatorStageRegistry.stageCount()];
+
 		Progress progress = new Progress(Integer.MAX_VALUE, 1000);
-		int numUpdated = 0;
+
 		while (!toUpdateSet.isEmpty()) {
-			numUpdated++;
 			GeneratorStage stage = processNext();
 			if (stage == null) {
-				skipped++;
+				++skipped;
```

Comment:
Changes like this are completely unnecessary. It isn't more efficient, and it doesn't seem more readable. In any case where it matters, the code should be changed to something more readable anyway.